File: DijkstraAlgorithm.py

""""

     ==> Step-by-Step Implementation <==

SOURCES: https://www.geeksforgeeks.org/dsa/dijkstras-shortest-path-algorithm-greedy-algo-7/

=> Set dist[source]=0 and all other distances as infinity.
=> Push the source node into the min heap as a pair <distance, node> → i.e., <0, source>.
=> Pop the top element (node with the smallest distance) from the min heap.
=> For each adjacent neighbor of the current node:
   * Calculate the distance using the formula:
        => dist[v] = dist[u] + weight[u][v]
    If this new distance is shorter than the current dist[v], update it.
        => Push the updated pair <dist[v], v> into the min heap
Repeat step 3 until the min heap is empty.
Return the distance array, which holds the shortest distance from the source to all nodes.
"""
import heapq
from collections import defaultdict
def Dijkstra(n,edges,src):
    graph=defaultdict(list)
    for s,t,w in edges:
        graph[s].append((t,w))
        graph[t].append((s,w)) #Remove if graph is directed
    
    dist=[float("inf")]*n
    dist[src]=0
    #Min-heap priority queue
    pq=[(0,src)]
    while pq:
        d,u=heapq.heappop(pq)
        #Skip Outdated entreis
        if d>dist[u]:
            continue
        
        for v,w in graph[u]:
            new_dist=dist[u]+w
            if new_dist<dist[v]:
                dist[v]=new_dist
                heapq.heappush(pq,(dist[v],v))
        
    
    return dist

n = 5
edges = [
    (0, 1, 2), (0, 2, 4),
    (1, 2, 1), (1, 3, 7),
    (2, 4, 3),
    (3, 4, 1)
]
src = 0

print(Dijkstra(n, edges, src))

File: Algorithms in cpp/DijkstraAlgorithm.cpp
//#include <bits/stdc++.h>
#include<iostream>
#include<vector>
#include<string>
#include <queue>

using namespace std;


vector<int> shortestReach(int n, vector<vector<int>> edges, int s)
{

    vector<vector<pair<int, int>>> graph(n);
    
    

    for (int i = 0; i < edges.size(); i++)
    {
        int nbr1 = edges[i][0]-1;
        int nbr2 = edges[i][1]-1;

        int weight = edges[i][2];
        graph[nbr1].push_back({nbr2, weight});
        graph[nbr2].push_back({nbr1, weight});
    }

    vector<int> dist(n, INT_MAX);
    dist[s - 1] = 0;

    // Min heap
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> heap;
    heap.push({0, s - 1});

    while (!heap.empty())
    {
         pair<int,int> topElement=heap.top();
        heap.pop();
        int currDist=topElement.first;
        int node=topElement.second;

        if (currDist > dist[node])
        {
            continue;
        }

        for (auto &edge : graph[node])
        {
            int nbr = edge.first;
            int weight = edge.second;
            int newDist = currDist + weight;
            if (newDist < dist[nbr])
            {
                dist[nbr] = newDist;
                heap.push({newDist, nbr});
            }
        }
    }

    vector<int> res;

    for (int i = 0; i < n; i++)
    {
        if (i != s - 1)
        {
            if (dist[i] == INT_MAX)
            {
                res.push_back(-1);
            }
            else
            {
                res.push_back(dist[i]);
            }
        }
    }
    return res;
}
int main()
{
    

   

}


/*

class Solution
{
public:
    vector<int> dijkstra(int V, vector<vector<int>> &edges, int src)
    {

        vector<vector<pair<int, int>>> graph(V);

        for (int i = 0; i < edges.size(); i++)
        {
            int nbr1 = edges[i][0];
            int nbr2 = edges[i][1];
            int weight = edges[i][2];

            graph[nbr1].push_back({nbr2, weight});
            graph[nbr2].push_back({nbr1, weight});
        }

        vector<int> dist(V, INT_MAX);
        dist[src] = 0;

        priority_queue<pair<int, int>> heap; // max-heap
        heap.push({0, src});

        while (!heap.empty())
        {

            pair<int, int> topEl = heap.top();
            heap.pop();

            int currDist = -topEl.first;
            int node = topEl.second;
            if (currDist > dist[node])
            {
                continue;
            }

            for (auto &edge : graph[node])
            {
                int nbr = edge.first;
                int weight = edge.second;
                int newDist = weight + currDist;
                if (newDist < dist[nbr])
                {
                    dist[nbr] = newDist;
                    heap.push({-newDist, nbr});
                }
            }
        }
        return dist;
    }
};

*/


